### PR TITLE
Hiberntate search

### PR DESCRIPTION
**Pull request by murtuza-ranapur**
Please re-read your article a day or so after you finished writing it and work through this checklist to polish the quality. This will save all of us a lot of time during review.

## Content
- [X] the article explains *why* and not only *how*
- [X] (half-)sentences that express a meaningful idea that is understandable by itself own are **bold** to allow quick scanning of the text

## Text
- [X] the text is conversational ("we" instead of "you", easy language)
- [X] there are no "walls of text" (short sentences, short paragraphs with a single idea)
- [X] the text is inclusive ("the developers" and "they" instead of "the developer" and "he/she")
- [X] the text uses active voice instead of passive voice
- [X] names are spelled consistently throughout the article (correct uppercase / lowercase)
- [X] file names and variables are highlighted as `code`

## Structure
- [X] the article has a short intro before the first headline
- [X] by reading the table of contents (TOC) alone, I get a sense of the "story" of the article (check the TOC in the preview after raising the PR or by starting the blog locally on your machine)
- [X] if using sub-headlines, there's at least two sub-headlines below a main-headline
- [X] headlines are consistent throughout the article (same level of abstraction, same style - if one headline starts with a verb in imperative form, the other headlines on the same level should probably, too)

## Code Examples
- [ ] code examples are introduced with a ":" and explained below the code example
- [X] code examples are formatted to avoid horizontal scrolling (manual line breaks and 2-spaces indentation)
- [X] the text has been checked with [Grammarly](https://grammarly.com) (the free tier is enough)

**Feel free to raise the PR even if the checklist hasn't been worked through, yet**, to take advantage of the preview that is generated for each pull request (see "deploy/netlify" under "Show all checks"). But please only flag it to be reviewed after going through the checklist above.

### CODE REVIEW

File: _posts/2020-09-16-hibernate-search.md [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+  @ContainedIn//->[4]
+  @OneToMany(mappedBy = "user")
+  private List<Post> post;
+}
+```
+
+Couple of interesting things to notice here. Let's start with the `index` property.
+
+1. `Index.NO` this indicates that property will be stored in the index document, but it won't be indexed. We won't be able
+to perform any search operation over it. You might be thinking "Why not simply remove `@Field` annotation?". And the answer is that we still need this field for Projection (Covered in the next section).
```
Comment (thombergs) [inline _posts/2020-09-16-hibernate-search.md:225]: Instead of saying "Covered in the next section", do this:

And the answer is that we still need this field for [Projection](Link to the section that explains projection)

File: _posts/2020-09-16-hibernate-search.md [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+you can see `@Field(name = "fullName")` is mapped to `first` and `last` both. This way `fullName` will have content of both fields. So, instead of searching over the `first` and `last` field separately, we can directly search over `fullName`. 
+
+3. We can set store to `Store.YES` when we plan to use it in projection. Note that this will require extra space. Plus, 
+Elasticsearch already stores the value in the `_soucre` field. So, the only reason to set the store to true is that when we don't want Elasticsearch to look up and extract value from the `_source` field.
+
+4. `@ContainedIn` just like `@OneToMany`-`@ManyToOne` annotation makes relationship bidirectional. When the values of this
+entity will be updated its values in the root entity i.e., in the `idx_post`(Post) index will also be updated.
+
+## Performing Queries
+Before we perform any queries, we first need to load data into the Elasticsearch.
```
Comment (thombergs) [inline _posts/2020-09-16-hibernate-search.md]: A sentence like "Before we do X" is a clear sign that this should be its own subsection.

File: _posts/2020-09-16-hibernate-search.md [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+      fullTextEntityManager.createIndexer().startAndWait();
+      log.info("All entities indexed");
+  }
+}
+```
+
+You can call the `initiateIndexing()` function either at the application startup or create an API in the REST controller to call it.
+`createIndexer` function also takes in Class references as input. This gives you more choice over the indexing. This is going to be a one-time thing, after this Hibernate search will keep entities in both sources in sync. Unless of course for some reason your database goes out of sync with Elasticsearch in which case indexing API might come in handy again. 
+
+With Elasticsearch integration you have two choices for writing queries:
```
Comment (thombergs) [inline _posts/2020-09-16-hibernate-search.md:264]: This would be the start for a second sub-section.

File: _posts/2020-09-16-hibernate-search.md [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+1. Hibernate search query DSL - It's a nice way to write Lucene queries. If you are familiar with Specification and
```
Comment (thombergs) [inline _posts/2020-09-16-hibernate-search.md]: Would be nice if Specification and Criteria would be linked to their docs.

File: _posts/2020-09-16-hibernate-search.md [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+    FullTextQuery fullTextQuery = fullTextEntityManager
+                          .createFullTextQuery(foodQuery, Post.class);
+    return (List<Post>) fullTextQuery.getResultList();
+  }
+  //.........
+  
+}
+```
+
+Let me explain the above code line by line:
```
Comment (thombergs) [inline _posts/2020-09-16-hibernate-search.md]: This sentence comes across as arrogant (I know it's not meant that way). I replaced it.

File: _posts/2020-09-16-hibernate-search.md [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+1. First, we create an object of `FullTextEntitiyManager` which is a wrapper over our `EntityManager`
+2. Next, we create `QueryBuilder` for the index on which we want to perform a search. We also need to pass the entity class object in it.
+3. Using `QueryBuilder` we finally build our `Query`. 
+4. Next, we make use of the keyword query `keyword()` which allows us to look for a specific word in a field or fields. Finally, we pass the word that we want to search in the `matching` 
+function.
+4. Lastly, we wrap everything in `FullTextQuery` and fetch result list by calling `getResultList()` function.
+
+One thing to note here is that although we are performing a query on Elasticsearch, Hibernate will still fire a query on the database to fetch the full entity. Which make sense, because as we saw in the previous section we didn't store all the fields of 
+`Post` entity and those fields still needed to be retrieved. If you only wanted to fetch what's stored in your index anyway and think this database call is redundant, I know just the way. Keep reading on to find out.
```
Comment (thombergs) [inline _posts/2020-09-16-hibernate-search.md]: The "I know just the way" is a bit too much :). Instead of "keep reading", and have the reader manually scan all the text, please put that piece of information into its own sub-section and link to that sub section here.

File: _posts/2020-09-16-hibernate-search.md [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+
+Let's move to the next query.
```
Comment (thombergs) [inline _posts/2020-09-16-hibernate-search.md:319]: A sentence like this makes me think this could be its own sub section. Can you think of a title for that subsection? What makes this query different from the first?

File: _posts/2020-09-16-hibernate-search.md [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+            .createQuery();
+
+  FullTextQuery fullTextQuery = fullTextEntityManager
+                .createFullTextQuery(finalQuery, Post.class);
+  fullTextQuery.setSort(qb.sort().byScore().createSort());
+  return (List<Post>) fullTextQuery.getResultList();
+}
+```
+
+As you can see in the snippet above for `likeCount` we are using range query. Using only `above()` is equivalent to `>=`
```
Comment (thombergs) [inline _posts/2020-09-16-hibernate-search.md]: "As you can see..." may turn some readers off if they don't see it at the first glance (which almost no one does, I bet). I changed the phrase.

File: _posts/2020-09-16-hibernate-search.md [hunk 9]
```diff
@@ -0,0 +1,440 @@
...
+As you can see in the snippet above for `likeCount` we are using range query. Using only `above()` is equivalent to `>=` 
+operator. If you want to exclude the limits, just call `excludeLimit()` after `above()`.
+  
+For the other two fields, we have again used a keyword query. 
+
+Now, it's time to combine all queries. To do so, we will make use of `QueryBuilder` object's `bool()` function which 
+provides us with verbs such as `should()`, `must()` and `not()`. I have used `must()` for `likeCount` query and `should()`
+for the rest as they are optional. Optional queries wrapped in `should()` contribute to the relevance score. For instance document with tag matching, 'Literature' will score higher.
+
+A couple of paragraphs ago I had promised you that I will show you a way to avoid queries to the database. So, here it is, and
```
Comment (thombergs) [inline _posts/2020-09-16-hibernate-search.md:378]: Yeah, no one will remember that :). Make this a separate section and link it above.

File: _posts/2020-09-16-hibernate-search.md [hunk 10]
```diff
@@ -0,0 +1,440 @@
...
+operator. If you want to exclude the limits, just call `excludeLimit()` after `above()`.
+  
+For the other two fields, we have again used a keyword query. 
+
+Now, it's time to combine all queries. To do so, we will make use of `QueryBuilder` object's `bool()` function which 
+provides us with verbs such as `should()`, `must()` and `not()`. I have used `must()` for `likeCount` query and `should()`
+for the rest as they are optional. Optional queries wrapped in `should()` contribute to the relevance score. For instance document with tag matching, 'Literature' will score higher.
+
+A couple of paragraphs ago I had promised you that I will show you a way to avoid queries to the database. So, here it is, and 
+it's called **Projection**.
```
Comment (thombergs) [inline _posts/2020-09-16-hibernate-search.md]: Makes me think that this should be its own section called "Projections" :)

File: _posts/2020-09-16-hibernate-search.md [hunk 11]
```diff
@@ -0,0 +1,440 @@
...
+```java
+public List<User> getUserByFirstWithProjection(String first,
```
Comment (thombergs) [inline _posts/2020-09-16-hibernate-search.md:382]: By the way, this code formatting is not optimal for code snippets in blog posts because it uses up too much whitespace. I replaced the indentation with 2 spaces everywhere.

File: _posts/2020-09-16-hibernate-search.md [hunk 12]
```diff
@@ -0,0 +1,440 @@
...
+      user.setAge((Integer) objects[4]);
+      users.add(user);
+  }
+  return users;
+}
+```
+
+Before moving on to the main event let's zoom in on a couple of interesting things here.
+
+* Up until now we were using keyword queries to perform exact match searches only, but here as you can see I have combined
```
Comment (thombergs) [inline _posts/2020-09-16-hibernate-search.md:443]: This is too much text for a list. A list means you can't separate the text into multiple paragraphs, so it becomes a wall of text. Please make this normal text and split it into meaningful paragraphs to reduce the wall of text.

File: _posts/2020-09-16-hibernate-search.md [hunk 13]
```diff
@@ -0,0 +1,440 @@
...
+it with the fuzzy query which will allow us to perform fuzzy searches too. This query gives end-users some flexibility in terms of searching by allowing some degree of error. The threshold of the error to be allowed can be decided by us. For instance, here I have
+set edit distance to 2 (Default is also 2 by the way) which means Elasticsearch will match all the words with a maximum of 2 differences to 
+the input. e.g., 'jab' will match 'jane'
+
+* We can also perform sorting and pagination. Sorting can be applied on single or multiple fields, and pagination can be 
+controlled with `setMaxResult`--Number of results we need per page-- and `setFirstResult`--We can pass page number in it.
```
Comment (thombergs) [inline _posts/2020-09-16-hibernate-search.md:449]: The explanation of setMaxResult and setFirstResult is not clear to me. Can you rephrase?

File: _posts/2020-09-16-hibernate-search.md (hunk 12 above)
Comment (thombergs) [reply _posts/2020-09-16-hibernate-search.md:443]: Actually, I think that "Fuzzy Search" and "Pagination" should each have their own subsection. Please think about creating a small code snippet for each, that focuses on each feature alone and creating a separate section for each. 

A section can very well be a short code snippet followed by a single paragraph. It doesn't need to be long

File: _posts/2020-09-16-hibernate-search.md [hunk 14]
```diff
@@ -0,0 +1,440 @@
...
+
+* Finally, let's talk about projection. To use projection we need to pass the list of fields that we want in output in the `setProjection` method.
```
Comment (thombergs) [inline _posts/2020-09-16-hibernate-search.md:451]: Above, you're saying that we're going to talk about projection, and now it's just a side note (the third bullet point in a list). Make this its own subsection and it will have more weight :).

File: _posts/2020-09-16-hibernate-search.md [hunk 15]
```diff
@@ -0,0 +1,440 @@
...
+controlled with `setMaxResult`--Number of results we need per page-- and `setFirstResult`--We can pass page number in it.
+
+* Finally, let's talk about projection. To use projection we need to pass the list of fields that we want in output in the `setProjection` method. 
+Now when we fetch results hibernate will return a list of object arrays. Which to be honest is a bit incontinent. Apart from
+fields, we can also fetch metadata such as id with `FullTextQuery.ID` or even score with `FullTextQuery.SCORE`.
+
+That's it! I mean this is not everything, but I believe this is enough to get you started. For further reading you can 
+explore the following:
+
+* Phrase queries - which allows you to search complete sentences
```
Comment (thombergs) [inline _posts/2020-09-16-hibernate-search.md:458]: Would be nice if you added links in this list, that makes a "Further Reading" section so much more valuable.

**Review by thombergs — CHANGES_REQUESTED** [review]
Hey!

good content! We can make this a really valuable article if we structure it a bit more. Basically, split the current sections into multiple subsections each to make the text more easily digestible and more searchable. I've left a few comments on that in the review. 

Looking forward to the end result!

File: _posts/2020-09-16-hibernate-search.md [hunk 16]
```diff
@@ -0,0 +1,440 @@
...
+* Further properties and its details can be found in Hibernate Search Document's 
+[Elasticsearch integration section](https://docs.jboss.org/hibernate/stable/search/reference/en-US/html_single/#elasticsearch-integration-configuration).
+
+
+One more thing to note here is that Hibernate Search v.5 only supports Elasticsearch up to v.5.2.x, though I have been using it with v.6.8, and it's working just fine. But, if you are using or planning on using Elasticsearch v.7 you might want to use Hibernate Search v.6 which is still in Beta by the way.
+
+If you choose to stick with Lucene which is the default integration you can still follow along as the APIs are almost 
+identical across integrations.
+
+If you have Spring boot actuator dependency in your project make sure to this property too
```
[the remark this answers is not in the document]
Comment (murtuza-ranapur) [reply _posts/2020-09-16-hibernate-search.md]: Hi @thombergs the Elasticsearch stats given out by the actuator are only of the cluster health. Should I go about explaining the fields in the section? 
```       
 "elasticsearchRest": {
            "status": "UP",
            "details": {
                "cluster_name": "docker-cluster",
                "status": "green",
                "timed_out": false,
                "number_of_nodes": 2,
                "number_of_data_nodes": 2,
                "active_primary_shards": 23,
                "active_shards": 46,
                "relocating_shards": 0,
                "initializing_shards": 0,
                "unassigned_shards": 0,
                "delayed_unassigned_shards": 0,
                "number_of_pending_tasks": 0,
                "number_of_in_flight_fetch": 0,
                "task_max_waiting_in_queue_millis": 0,
                "active_shards_percent_as_number": 100.0
            }
        }
```

File: _posts/2020-09-16-hibernate-search.md [hunk 17]
```diff
@@ -0,0 +1,556 @@
...
+`@Field(index = Index.NO, store = Store.YES)` - `Index.NO` indicates that field won't be indexed. We won't be able to perform any search operation over it. You might be
+thinking "Why not simply remove `@Field` annotation?". And the answer is that we still need this field for [Projection](#projection).
+
+#### Combine Field Data
+`@Field(name = "fullName")` - In the `Post` class we saw that we can map one entity field to multiple index document fields. We can also do the inverse. As 
+you can see `@Field(name = "fullName")` is mapped to `first` and `last` both. This way `fullName` will have the content of both fields. So, instead of searching over the `first` and `last` field separately, we can directly search over `fullName`. 
+
+#### Store Property
+`@Field(store = Store.YES)` - We can set store to `Store.YES` when we plan to use it in projection. Note that this will require extra space. Plus, 
+Elasticsearch already stores the value in the `_source` field. So, the only reason to set the store to true is that when we don't want Elasticsearch to look up and extract value from the `_source` field.
```
Comment (thombergs) [inline _posts/2020-09-16-hibernate-search.md:272]: What is the `_source` field? Please explain or don't mention it.

File: _posts/2020-09-16-hibernate-search.md [hunk 18]
```diff
@@ -0,0 +1,556 @@
...
+Elasticsearch indexes.
+
+To perform pagination we need two things, the number of results we want per page and page offset (Page number, if I put plainly).
+Prior we can pass to `FullTextQuery` objects `setMaxResult` method and later in `setFirstResult`. Then query will return results 
+accordingly.
+
+Query DSL also provides us a way to define sort field and order using `sort()`. We can also perform sort operation on multiple fields by 
+chaining with `andByField()`.
+
+That's it! I mean this is not everything, but I believe this is enough to get you started. For further reading you can
```
Comment (thombergs) [inline _posts/2020-09-16-hibernate-search.md:555]: This and below shouldn't be part of the Pagination section. You can call it "Further Reading" or something similar.

File: _posts/2020-09-16-hibernate-search.md [hunk 19]
```diff
@@ -0,0 +1,556 @@
+---
+title: "Full-Text Search with Hibernate Search and Spring Boot"
```
Comment (thombergs) [inline _posts/2020-09-16-hibernate-search.md:2]: We'll have to think of a better title because there is no Spring Boot in the article.

**Review by thombergs — CHANGES_REQUESTED** [review]
It reads much better now that it's split up in sections. I've pushed some changes to your branch.

Please go through my comments one more time and address them. The main issue is the big code examples which I think should be cut up and distributed across the sections so that the reader doesn't have to go back-and-forth.

In the code-examples PR, please add your module to the `build-all.sh` file so that it will be part of the CI build.
